### PR TITLE
Login Functionality & Multiple pages in App & Allow Choosing skills & Developer View

### PR DESCRIPTION
**Pull request by cs1090218**
- User Registration and Login via Cognito (logout functionality still remains + some security hooks)
- We have multiple pages in the App, one for main page and another for Evaluation UI
- Allows choosing skill from a dropdown menu so we don't have to restart server
- There's a developer view which shows some additional features compared to student view

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -6,25 +6,121 @@
...
+REDIRECT_URL = (
+    "http://localhost:8000/ask-ivy" if ON_LOCALHOST else "https://dev.dilab-ivy.com"
+)
+CLIENT_ID = (
+    "60p8a9bvteiihrd8g89r5ggabi" if ON_LOCALHOST else "2d7ah9kttong2hdlt4olhtao4d"
+)
+CLIENT_SECRET = (
+    os.getenv("COGNITO_LOCALHOST_CLIENT_SECRET")
+    if ON_LOCALHOST
+    else os.get_env("COGNITO_PROD_CLIENT_SECRET")
```
Comment (gracebrazil28) [inline main.py:45]: Should this also be os.getenv?

File: main.py [hunk 2]
```diff
@@ -6,25 +6,121 @@
...
+LOGIN_URL = (
+    COGNITO_DOMAIN
+    + f"/login?response_type=code&client_id={CLIENT_ID}&redirect_uri={REDIRECT_URL}"
+)
+GET_ACCESS_TOKEN_URL = COGNITO_DOMAIN + "/oauth2/token"
+GET_USER_INFO_URL = COGNITO_DOMAIN + "/oauth2/userInfo"
+
+USER_NAME = ""
+USERNAME = ""
+USER_EMAIL = ""
```
Comment (gracebrazil28) [inline main.py:56]: Should we try to implement maybe using a class for managing user info, avoiding using global variables? 

class UserInfo:
    def __init__(self):
        self.name = ""
        self.username = ""
        self.email = ""

File: main.py [hunk 3]
```diff
@@ -237,6 +352,15 @@ def handle_download_click():
...
+# Launch the Application
+ivy_main_page.queue()
+
+with gr.Blocks() as evaluation_page:
+    # Develop Evaluation page
+    # Placeholder interface
+    gr.Interface(lambda x: f"Hello {x}!", inputs="text", outputs="text")
+
+app = gr.mount_gradio_app(
+    app, ivy_main_page, path="/ask-ivy", allowed_paths=["flagged/", "commended/"]
```
Comment (gracebrazil28) [inline main.py:364]: Do we still need the flagged and commended paths?

**Comment by gracebrazil28** [comment]
I wanted to try this on my local environment, was this tested? I'm running into dependency issues in botocore:
`aiobotocore 2.12.3 requires botocore<1.34.70,>=1.34.41, but you have botocore 1.34.149 which is incompatible.
`
Maybe we can update it also. At that point we can circle back to the duration limit for gr.Info.

File: main.py (hunk 1 above)
Comment (cs1090218) [reply main.py:45]: Added to coolify.

File: main.py (hunk 2 above)
Comment (cs1090218) [reply main.py:56]: Yes that is a good idea. This file is becoming unmanageable very quickly.
One of us should take a refactor task for next sprint.

File: main.py (hunk 3 above)
Comment (cs1090218) [reply main.py:364]: Good point, adding a commit to remove it.

**Review by gracebrazil28 — COMMENTED** [review]
This download button is causing issues: 
image

Addressing this in my next commit.

**Comment by gracebrazil28** [comment]
> This download button is causing issues: image
> 
> Addressing this in my next commit.

Oh I see, PR #3 changes in line 88 is not reflected here. That's why there's no download btn functionality yet. Let's merge this and then resolve conflicts. When I changed this to this in my local development it worked: gr.Button -> gr.DownloadButton
```
download_btn = gr.DownloadButton(
            label="Download Flagged Responses",
            variant="secondary",
            visible=IS_DEVELOPER_VIEW,
        )
```

I'm gonna let it merge and then we can test locally again. I just added a new test script in /test_scripts to quickly check communication between cognito and web app.

I was also able to verify the cognito functionality (I am user grace_test). Great work!

**Review by gracebrazil28 — APPROVED** [review]

**Review by gracebrazil28 — APPROVED** [review]

**Pull request merged**